frontend/frontend.py

```diff
@@ -47,10 +46,13 @@ def draw_gradio_ui(opt, img2img=lambda x: x, txt2img=lambda x: x, txt2img_defaul
                                 gr.Markdown(
                                     'Select an image from the gallery, then click one of the buttons below to perform an action.')
                                 with gr.Row():
-                                    output_txt2img_copy_clipboard = gr.Button("Copy to clipboard").click(fn=None,
-                                                                                                         inputs=output_txt2img_gallery,
-                                                                                                         outputs=[],
-                                                                                                         _js=js_copy_to_clipboard('txt2img_gallery_output'))
+                                    output_txt2img_copy_clipboard = gr.Button("Copy to clipboard")\
+                                                                      .click(fn=None,
+                                                                             inputs=output_txt2img_gallery,
+                                                                             outputs=[],
+                                                                             _js=call_SD("copyImageFromGalleryToClipboard",
+                                                                                         fromId="'txt2img_gallery_output'")
```

Comment:
Can we do something about this double string requirement in fromId?

Reply:
Related: https://github.com/hlky/stable-diffusion-webui/pull/381#discussion_r959814261

This could be handled here ^ if we accept a dictionary for arguments and convert it into a JSON string.

Reply:
We can assume that every argument is a string but I really wish to retain the ability to pass other types.
See line 70 where I pass an object instead

Reply:
@codedealer 

I would do it like this:

```py
import json

some_options = {
  "num": 123,
  "bool": True,
  "str": "this is a string"
}

json_str = json.dumps(some_options)

function_name = "some_function"

function_call_str = f"async (x) => {{ return await SD.{function_name}({{ ...{json_str}, x }}) ?? [] }}"
```

Reply:
Types are useful.

Reply:
@nderscore I phrased this poorly, in fact I'm not sure how to describe it so I'll provide a real example:

 How do I get a reference to a js object in this way? dict passed in `json.dumps()` requires all arguments to be valid python values.
Currently I have this
`call_SD("textToClipboard", text="x")`
where x references _the js object that does not exist during python's execution_. With proposed way this `"x"` argument will be treated as a string by JSON and it will subsequently pass a string to your js method when you wanted an object reference.

Reply:
In my example, the `x` is still being proxy-passed through like before.

Shoudn't `SD.textToClipboard` read that value from `{ x }` instead?

```py
# text to clipboard only uses `x` from gradio, so it doesn't need any custom arguments
call_SD("textToClipboard")
```
```js
// no options should be an empty object:
async (x) => { return await SD.textToClipboard({ ...{}, x }) ?? []; }
// which is interpretted like this:
async (x) => { return await SD.textToClipboard({ x })
// and then the textToClipboard method should use x:
textToClipboard({ x }) {
  // copy x to clipboard
}
```

Reply:
Maybe doing things like `text="x"` is not such a great idea? 🤔

Reply:
@nderscore I tried to keep `textToClipboard()` argument agnostic in case we use it in the future from js itself. But yeah, technically speaking it doesn't need any arguments right now.

Reply:
You could have two functions: One that's a generic utility function that takes a string, and one that's a wrapper around that utility function which consumes `x` from gradio

Reply:
working on it as we speak!